### ISSUE TITLE
METIS Partitioning help

### ISSUE THREAD

**Issue by Jianming-Pu** [GEOS-DEV/GEOS#2915]
**Describe the question**
I want to add element weight information when initially dividing the domain using METIS, but after re-division the calculation result is correct for solids and incorrect for fluid-solid coupling.
My method is as follows: 

1. divide all elements into Rank0 to get the cell's constitutive information and the weight information
2. re-divide domain with METIS

**Platform**

1. GEOSX version 0.2.0 ( sha1: 42d058a41)
3. Machine [Ubuntu18]
4. Compiler: [gcc 8.3.0]

**Comment by klevzoff** [comment]
Hi @Jianming-Pu,

Is this approach something you've already tried to implement in GEOS? If so, would you mind sharing the code by opening a PR? Currently in GEOS mesh (re-)distribution is done in isolation from any constitutive or physics information, so I'm curious to see how you're incorporating that information into the partitioning. 

An "easy" way to include weights would be to precompute them in mesh pre-processing workflow (either in Paraview or via Python scripts) and store as a cell field in the VTK dataset. Then a fairly trivial modification inside `VTKMesh` generator would enable it to consume the weights. On the other hand, if you would like the weights to be computed by GEOS automatically based on assignment of target regions to different physics solvers, that is a much more significant change, and I'm not sure there exists a universally accepted weight assignment scheme.

In short, could you please elaborate a bit more on
1) what you are trying to achieve
2) what you have already tried

and we can discuss potential solutions.

Thanks!

**Comment by Jianming-Pu** [comment]
Thanks for your reply.
I'm sorry I'm building from a source tree that is not a git repository, and it's an older version. But I can give a general idea of the changes I'm making.

My goal is to implement metis partitioning based on the weights of the material partitions. This weight is written as a parameter of the material in the `xml` document.

My main modification is in the `ProblemManager::problemSetup()` function, which executes twice `generateMesh` and `applyNumericalMethods`. 

1. In the first call to metis by simple processing so that all the elements are divided in Rank0, and then after the `hangConstitutiveRelation` function in `applyNumericalMethods `we can use `getReference` to get the elements' constitutive  information
2. Then executing` generateMesh` and `applyNumericalMethods` once more realizes the redistribution.

But after verifying that the solids are correct and the fluid-solid coupling is incorrect, I tested and found that the problem is in the second `hangConstitutiveRelation` function, which I skipped the second time because the first time I had already registered the constitutive information. So maybe I should deregister it and re-execute the hangConstitutiveRelation function again?
